File: curricula/systems/kernel/tools/verify.py
#!/usr/bin/env python3
"""فاحصُ منهج الكيرنل.

يقرأ `regions/*.md` ويفرض على كل لوحةٍ أن تكون منسوبةً ومُتحقَّقاً منها:

  <!-- runs: NN -->     اللوحة مخرَجُ تشغيلِ المرحلة NN — تُبنى وتُقلَع وتُقارَن
  <!-- part: PATH -->    البلوك مقتطعٌ من ملفٍّ حقيقيّ — تُفحَص أسطرُه فيه
  <!-- out @tag: … -->   وسمُ السلطة إلزاميّ لكل لوحة مخرَج

والمقارنة «تضمُّنٌ بالترتيب»: كل سطرٍ غير مطويٍّ في اللوحة يجب أن يظهر في المخرَج
الحقيقيّ بنفس الترتيب. فاللوحةُ مقتطعٌ مشروع، والاختلاقُ ليس كذلك.

وما لا يُفحَص بالتشغيل — نصُّ مواصفةٍ، أو قياسٌ على جهاز القارئ — يُعَدّ ويُذكَر
في التقرير. الثقبُ المعلوم أهونُ من الصامت.
"""
import os, re, subprocess, sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
REGIONS = os.path.join(ROOT, "regions")
PROGRAMS = os.path.join(ROOT, "programs")

TAGS = {"arch", "abi", "boot", "qemu", "ours", "fw", "unspec", "host"}
# وسومٌ لا يفحصها التشغيل، ومعها سببُ ذلك
NOT_RUN = {"host": "قياسٌ على جهاز القارئ", "abi": "بنيةُ ملفٍّ تُقرأ بالأدوات",
           "fw": "خريطةٌ من الـfirmware", "unspec": "غيرُ محدَّدٍ معمارياً"}
VARY = re.compile(r"تختلف|يختلف|تتغيّر|يتغيّر")

err, warn = [], []
def E(m): err.append(m)
def W(m): warn.append(m)

def norm(s, vary):
    s = re.sub(r"\s+", " ", s.strip())
    if vary:
        s = re.sub(r"0x[0-9a-fA-F]+", "#", s)
        s = re.sub(r"\b[0-9a-f]{6,16}\b", "#", s)
        s = re.sub(r"\d+", "#", s)   # أُعلن التغيّر، فيبقى الشكلُ والترتيب لا القيم
    return s

_cache = {}
def boot(stage):
    """يبني المرحلة ويقلعها ويعيد طرفيتها."""
    if stage in _cache: return _cache[stage]
    build = os.path.join(PROGRAMS, "build.sh")
    try:
        r = subprocess.run([build, stage, "run"], capture_output=True, text=True, timeout=300)
        out = r.stdout
    except Exception as e:
        out = ""
        E(f"المرحلة {stage}: تعذّر التشغيل — {e}")
    _cache[stage] = out
    return out

def parse(path):
    """يعيد قائمة اللوحات: (نوع, وسم, ملاحظة, أسطر, سطرُ البداية)."""
    lines = open(path, encoding="utf-8").read().split("\n")
    panels, pend_run, pend_part = [], None, None
    i = 0
    while i < len(lines):
        ln = lines[i]
        m = re.match(r"<!--\s*runs:\s*(\S+)\s*-->", ln)
        if m: pend_run = m.group(1); i += 1; continue
        m = re.match(r"<!--\s*part:\s*(\S+)\s*-->", ln)
        if m: pend_part = m.group(1); i += 1; continue
        m = re.match(r"<!--\s*out\s+@(\w+)\s*:\s*(.*?)\s*-->", ln)
        if m:
            tag, note = m.group(1), m.group(2)
            j = i + 1
            while j < len(lines) and lines[j].strip() == "": j += 1
            if j >= len(lines) or not lines[j].startswith("```"):
                E(f"{os.path.basename(path)}:{i+1}: وسمُ out بلا كتلةٍ بعده"); i += 1; continue
            k = j + 1; body = []
            while k < len(lines) and not lines[k].startswith("```"): body.append(lines[k]); k += 1
            panels.append(("out", tag, note, body, i + 1, pend_run))
            pend_run = None
            i = k + 1; continue
        if ln.startswith("```") and pend_part:
            k = i + 1; body = []
            while k < len(lines) and not lines[k].startswith("```"): body.append(lines[k]); k += 1
            panels.append(("part", pend_part, "", body, i + 1, None))
            pend_part = None
            i = k + 1; continue
        if re.match(r"<!--\s*out\b", ln):
            E(f"{os.path.basename(path)}:{i+1}: لوحةُ مخرَجٍ بلا وسم سلطة")
        i += 1
    return panels

def check_part(path, src, body, at):
    f = os.path.join(PROGRAMS, src)
    if not os.path.exists(f):
        E(f"{os.path.basename(path)}:{at}: part يشير إلى ملفٍّ غير موجود → {src}")
        return
    have = [re.sub(r"\s+", " ", l.strip()) for l in open(f, encoding="utf-8")]
    missing = 0
    for l in body:
        t = re.sub(r"\s+", " ", l.strip())
        if not t or t.startswith("/*") or t.startswith("*") or t in ("...", "…"): continue
        if t not in have: missing += 1
    if missing:
        E(f"{os.path.basename(path)}:{at}: {missing} سطراً في المقتطع ليست في {src}")

def check_out(path, tag, note, body, at, stage):
    if tag not in TAGS:
        E(f"{os.path.basename(path)}:{at}: وسمٌ غير معروف @{tag}")
        return None
    if not note:
        E(f"{os.path.basename(path)}:{at}: وسم @{tag} بلا ملاحظةٍ تقول مستوى الرصد")
    if stage is None:
        return ("unrun", tag)
    real = boot(stage)
    if not real:
        E(f"{os.path.basename(path)}:{at}: المرحلة {stage} لم تُنتج مخرَجاً")
        return ("fail", tag)
    vary = bool(VARY.search(note))
    hay = [norm(l, vary) for l in real.split("\n")]
    pos, miss = 0, []
    for l in body:
        t = l.strip()
        if not t or t in ("...", "…") or t.startswith("$"): continue
        n = norm(l, vary)
        try:
            pos = hay.index(n, pos) + 1
        except ValueError:
            miss.append(t[:70])
    if miss:
        E(f"{os.path.basename(path)}:{at}: {len(miss)} سطراً لا يظهر في تشغيل {stage} — أوّلُها «{miss[0]}»")
        return ("fail", tag)
    return ("ok", tag)

def main():
    only = sys.argv[1] if len(sys.argv) > 1 else None
    stats = {"ok": 0, "unrun": 0, "fail": 0, "part": 0}
    unrun_by_tag = {}
    files = sorted(f for f in os.listdir(REGIONS) if f.endswith(".md"))
    for f in files:
        if only and only not in f: continue
        p = os.path.join(REGIONS, f)
        for kind, a, note, body, at, stage in parse(p):
            if kind == "part":
                check_part(p, a, body, at); stats["part"] += 1
            else:
                r = check_out(p, a, note, body, at, stage)
                if r:
                    stats[r[0]] += 1
                    if r[0] == "unrun": unrun_by_tag[r[1]] = unrun_by_tag.get(r[1], 0) + 1

    print(f"مقتطعات مفحوصة في ملفّاتها: {stats['part']}")
    print(f"لوحات مقارَنةٌ بتشغيلٍ حقيقيّ:  {stats['ok']}")
    print(f"لوحات لم تُفحَص بالتشغيل:      {stats['unrun']}")
    for t, n in sorted(unrun_by_tag.items()):
        print(f"    @{t:7s} {n:2d}  — {NOT_RUN.get(t, 'يحتاج ربطاً بمرحلة')}")
    for m in warn: print("⚠ " + m)
    if err:
        print(f"\n✗ {len(err)} خطأ:")
        for m in err: print("  " + m)
        return 1
    print("\n✓ كل لوحةٍ منسوبةٌ، وكل مقتطعٍ من ملفّه، وكل مربوطٍ بمرحلةٍ يطابق تشغيلها")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: curricula/systems/kernel/site/scripts/verify.mjs
#!/usr/bin/env node
/**
 * فحص المحتوى — يقرأ `regions/` و`appendix/` و`README.md` و`programs/`
 * مباشرةً، ويفشل بالملفّ والسطر.
 *
 * وما يقرؤه `tools/verify.py` يقرؤه هذا بنفس الصيغة، فلا ينحرف العرض عمّا
 * يُتحقَّق منه.
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const CUR = path.resolve(HERE, '..', '..');
const REG = path.join(CUR, 'regions');
const AP = path.join(CUR, 'appendix');
const PROG = path.join(CUR, 'programs');
const SRC = path.resolve(HERE, '..', 'src');

let bad = 0;
const err = (m) => { console.error(`  ✗ ${m}`); bad++; };
const say = (m) => console.log(`  · ${m}`);

/* ── الصيغة، بنفس تعابير `tools/verify.py` ── */
const TAGS = ['arch', 'abi', 'boot', 'qemu', 'ours', 'fw', 'unspec', 'host'];
const FOLD = { arch: 'spec', abi: 'spec', boot: 'spec', qemu: 'env', fw: 'env', host: 'env', ours: 'ours', unspec: 'none' };
const RUNS = /^<!--\s*runs:\s*(\S+)\s*-->$/;
const PART = /^<!--\s*part:\s*(\S+)\s*-->$/;
const OUT = /^<!--\s*out\s+@(\w+)\s*:\s*(.*?)\s*-->$/;

/* مراحل القالب، بنفس ترتيب الأسلوب `kernel-dfs` §٤ */
const TEMPLATE = [
  'المشكلة', 'الصورة العامة', 'الآلية الخام', 'البنية الداخلية', 'مسار التنفيذ',
  'أصغر implementation', 'التجربة', 'كسرها عمداً', 'كيف تفعلها الأنظمة الحقيقية',
  'الفخاخ التصوّرية', 'Checkpoint', 'Artifact', 'التمرين',
];
const stageOf = (t) => TEMPLATE.find((s) => t === s || t.startsWith(s + ':')) ?? null;
const plain = (s) => s.replace(/`([^`]*)`/g, '$1').replace(/\*\*([^*]*)\*\*/g, '$1').trim();

const files = fs.readdirSync(REG).filter((f) => f.endsWith('.md')).sort();
const nums = files.map((f) => f.slice(0, 2));

/* ── أ) جدول README مقابل ما في regions/ ── */
const readme = fs.readFileSync(path.join(CUR, 'README.md'), 'utf8');
const planned = [...readme.matchAll(/^\|\s*\[`(\d\d)`\]/gm)].map((m) => m[1]);
const missing = planned.filter((n) => !nums.includes(n));
const extra = nums.filter((n) => !planned.includes(n));
if (missing.length) err(`فصولٌ في جدول README بلا ملفّ: ${missing.join(' ')}`);
if (extra.length) err(`ملفّاتٌ في regions/ بلا صفٍّ في جدول README: ${extra.join(' ')}`);
say(`${nums.length} فصلاً، وجدولُ README يطابقها`);

/* ── ب) المراحل من `programs/stages.conf` ── */
const conf = fs.readFileSync(path.join(PROG, 'stages.conf'), 'utf8');
const stages = conf.split('\n').map((l) => l.trim()).filter((l) => l && !l.startsWith('#'))
  .map((l) => l.split('|').map((x) => x.trim()));
const stageIds = new Set(stages.map((s) => s[0]));
const stageFiles = Object.fromEntries(stages.map((s) => [s[0], (s[1] ?? '').split(/\s+/).filter(Boolean)]));
say(`${stageIds.size} مرحلةً في stages.conf، منها ${[...stageIds].filter((s) => /x$/.test(s)).length} كسرٌ متعمَّد`);

for (const num of nums) {
  if (!stageIds.has(num) && num !== '00') err(`الفصل ${num}: لا مرحلة بهذا الرقم في stages.conf`);
}
/* ملفّ المرحلة إمّا تحت `kernel/` وإمّا مسارٌ صريحٌ تحت `programs/` — كـ`variants/` */
const progFile = (f) => fs.existsSync(path.join(PROG, 'kernel', f)) || fs.existsSync(path.join(PROG, f));
for (const [id, fl] of Object.entries(stageFiles)) {
  for (const f of fl) if (!progFile(f)) err(`stages.conf: المرحلة ${id} تذكر ملفّاً غير موجود — ${f}`);
}

/* ── ج) اللوحات والمقتطعات وأقسام القالب ── */
const fams = { spec: 0, env: 0, ours: 0, none: 0 };
let panels = 0, parts = 0, traces = 0, limits = 0, offTemplate = 0, subs = 0;
const brokenPanels = [];
const workingByCh = {};

for (const f of files) {
  const num = f.slice(0, 2);
  const lines = fs.readFileSync(path.join(REG, f), 'utf8').split('\n');
  let pendRun = null, pendPart = null;
  workingByCh[num] = new Set();

  lines.forEach((raw, i) => {
    const ln = raw.trim();
    const at = `${f}:${i + 1}`;

    let m = ln.match(RUNS);
    if (m) {
      pendRun = m[1];
      if (!stageIds.has(pendRun)) err(`${at}: runs يشير إلى مرحلةٍ ليست في stages.conf — ${pendRun}`);
      return;
    }
    m = ln.match(PART);
    if (m) {
      pendPart = m[1];
      parts++;
      if (!fs.existsSync(path.join(PROG, pendPart))) err(`${at}: part يشير إلى ملفٍّ غير موجود — ${pendPart}`);
      return;
    }
    m = ln.match(OUT);
    if (m) {
      const [, tag, note] = m;
      panels++;
      if (!TAGS.includes(tag)) err(`${at}: وسمٌ غير معروف @${tag}`);
      else fams[FOLD[tag]]++;
      if (!note) err(`${at}: وسم @${tag} بلا ملاحظةٍ تقول مستوى الرصد`);
      let j = i + 1;
      while (j < lines.length && !lines[j].trim()) j++;
      if (!lines[j]?.startsWith('```')) err(`${at}: وسمُ out بلا كتلةٍ بعده`);
      if (pendRun) {
        if (/x$/.test(pendRun)) brokenPanels.push({ ch: num, stage: pendRun, at });
        else workingByCh[num].add(pendRun);
      }
      pendRun = null;
      return;
    }
    if (/^<!--\s*out\b/.test(ln)) err(`${at}: لوحةُ مخرَجٍ بلا وسم سلطة`);
    if (ln.startsWith('⚠')) limits++;

    const h2 = raw.match(/^##\s+(.+)$/);
    if (h2 && !stageOf(plain(h2[1]))) offTemplate++;
    if (/^###\s+/.test(raw)) subs++;
  });

  /* كتلةُ مسار التنفيذ: بلا لغة وفيها `↓` — تُعدّ بمسحٍ سطريّ لا بتعبيرٍ يتداخل */
  let open = null, buf = [];
  for (const raw of lines) {
    if (raw.startsWith('```')) {
      if (open === null) { open = raw.slice(3).trim(); buf = []; }
      else { if (!open && buf.some((l) => l.includes('↓'))) traces++; open = null; }
    } else if (open !== null) buf.push(raw);
  }
}

say(`${panels} لوحةَ مخرَج · ${parts} مقتطعاً من ملفّه · ${traces} مسارَ تنفيذٍ · ${limits} حدَّ ادّعاء`);
say(`العائلات — المواصفة ${fams.spec} · هذه البيئة ${fams.env} · قرارُنا ${fams.ours} · بلا مالك ${fams.none}`);
say(`${offTemplate} قسماً خارج قالب الاثنتي عشرة مرحلة، و${subs} عنواناً فرعياً`);

/*
 * ── د) زوج «تعمل / كُسِرت» ──
 *
 * الزوج يُعرَض حين توجد **لوحةٌ عاملةٌ واحدة بالضبط** للمرحلة الأصل في المنهج —
 * لا صفر ولا اثنتان. وما لا مقابل له يُعَدّ ويُذكَر: العرض يسقط إلى لوحةٍ
 * واحدة، والثقبُ المعلوم أهونُ من الصامت.
 */
const allWorking = Object.values(workingByCh).flatMap((s) => [...s]);
let paired = 0;
const lone = [];
for (const b of brokenPanels) {
  const base = b.stage.slice(0, -1);
  const n = allWorking.filter((x) => x === base).length;
  if (n === 1) paired++;
  else if (n === 0) lone.push(b.stage);
  else err(`${b.at}: مرحلةُ الكسر ${b.stage} تقابلها ${n} لوحاتٍ عاملة — الزوج يحتاج واحدةً`);
}
say(`${brokenPanels.length} لوحةَ كسرٍ متعمَّد: ${paired} مزدوجةٌ بلوحتها العاملة${lone.length ? ` · ${lone.join(' ')} بلا مقابلٍ فتُعرَض وحدها` : ''}`);

/* ── هـ) أقسامٌ منتظمة: الفخاخ عمودان، وArtifact خمسةُ صفوف ── */
const ART = ['مخطّط', 'تنفيذ', 'تجربةُ فشل', 'أثرُ تنفيذ', 'مقارنة'];
let arts = 0, traps = 0, cps = 0;
for (const f of files) {
  const s = fs.readFileSync(path.join(REG, f), 'utf8');
  const grab = (h) => {
    const m = s.match(new RegExp(`^## ${h}\\s*$\\n([\\s\\S]*?)(?=^## |\\Z)`, 'm'));
    return m ? m[1] : null;
  };
  const a = grab('Artifact');
  if (a) {
    arts++;
    const labs = a.split('\n').filter((l) => l.trim().startsWith('|')).slice(2)
      .map((l) => l.split('|')[1].trim()).filter(Boolean);
    if (JSON.stringify(labs) !== JSON.stringify(ART)) err(`${f}: صفوف Artifact ليست الخمسة المعلَنة — ${labs.join(' · ')}`);
  }
  const t = grab('الفخاخ التصوّرية');
  if (t) {
    traps++;
    const head = t.split('\n').find((l) => l.trim().startsWith('|'));
    const cols = head.split('|').slice(1, -1).map((c) => c.trim());
    if (cols[0] !== 'الفخّ' || cols[1] !== 'الصواب') err(`${f}: جدول الفخاخ بعمودين غير «الفخّ» و«الصواب»`);
  }
  const c = grab('Checkpoint');
  if (c) {
    cps++;
    const qs = (c.match(/^[٠-٩]+\.\s/gm) ?? []).length;
    if (qs < 2) err(`${f}: Checkpoint بأقلّ من سؤالين`);
  }
}
say(`${arts} جدول Artifact · ${traps} جدول فخاخ · ${cps} بوّابة Checkpoint`);

/* ── و) الملاحق ── */
const aps = fs.readdirSync(AP).filter((f) => f.endsWith('.md'));
say(`${aps.length} ملحقاً`);

/* ── ز) ثنائية اللغة في مصدر الموقع ── */
const walk = (d, out = []) => {
  for (const e of fs.readdirSync(d, { withFileTypes: true })) {
    const p = path.join(d, e.name);
    if (e.isDirectory()) walk(p, out);
    else out.push(p);
  }
  return out;
};
const srcFiles = walk(SRC);
const ARABIC = /[؀-ۿ]/;

for (const p of srcFiles.filter((x) => /\.tsx?$/.test(x))) {
  const s = fs.readFileSync(p, 'utf8');
  s.split('\n').forEach((ln, i) => {
    if (/<text[ >]/.test(ln) && ARABIC.test(ln) && !/text-anchor|textAnchor/.test(ln))
      err(`${path.relative(SRC, p)}:${i + 1}: <text> عربيّ بلا text-anchor`);
  });
}

/*
 * ولا نصَّ عربيٍّ مكتوبٍ في مكوّنٍ بصنفٍ أحاديّ.
 *
 * وهذا فحصُ **مصدرٍ** لا تصيير، لأن التصيير لا يمرّ على كل حال: وسمُ «قبل فتح
 * المصدر» في `Trail.tsx` لا يظهر إلا بعد أن يكتب القارئ توقّعه، فلا يراه
 * `ssr-check` أصلاً — ورآه هذا.
 */
const MONO_CLASSES = (() => {
  const out = new Set();
  const all = ['components.css', 'base.css', 'layout.css']
    .map((f) => fs.readFileSync(path.join(SRC, 'styles', f), 'utf8')).join('\n');
  for (const m of all.matchAll(/([^{}]+)\{([^}]*)\}/g)) {
    if (!/--k-mono/.test(m[2])) continue;
    for (const c of m[1].matchAll(/\.([\w-]+)/g)) out.add(c[1]);
  }
  return out;
})();

for (const p of srcFiles.filter((x) => /\.tsx$/.test(x))) {
  fs.readFileSync(p, 'utf8').split('\n').forEach((ln, i) => {
    for (const m of ln.matchAll(/className="([^"]*)"[^>]*>([^<{]*)/g)) {
      const cls = m[1].split(/\s+/).find((c) => MONO_CLASSES.has(c));
      if (cls && ARABIC.test(m[2])) err(`${path.relative(SRC, p)}:${i + 1}: نصٌّ عربيّ في صنفٍ أحاديّ .${cls}`);
    }
  });
}

/* المونو للكود والأرقام فقط — كل مُحدِّدٍ يستعمله لا يحمل عربية */
const css = fs.readFileSync(path.join(SRC, 'styles/components.css'), 'utf8')
  + fs.readFileSync(path.join(SRC, 'styles/base.css'), 'utf8')
  + fs.readFileSync(path.join(SRC, 'styles/layout.css'), 'utf8');
const monoSel = [];
const missingIso = [];
for (const m of css.matchAll(/([^{}]+)\{([^}]*)\}/g)) {
  const sel = m[1].trim().replace(/\s+/g, ' ');
  const body = m[2];
  if (!/--k-mono/.test(body)) continue;
  monoSel.push(sel);
  /* `pre` و`code` معزولان في `base.css` أصلاً، وما عداهما يعزل نفسه */
  if (!/unicode-bidi|direction/.test(body) && !/\bpre\b|\bcode\b/.test(sel)) missingIso.push(sel);
}
if (missingIso.length) err(`محدِّدٌ بخطٍّ أحاديّ بلا عزل اتجاه: ${missingIso.join(' · ')}`);
say(`${monoSel.length} مُحدِّداً بالخطّ الأحاديّ، وكلُّها معزولةُ الاتجاه`);

if (/letter-spacing/.test(css)) err('تباعد حروف في أنماط الموقع — يقطع العربية');

console.log(bad ? `\n✗ ${bad} خطأً` : '\n✓ المحتوى والموقع متطابقان');
process.exit(bad ? 1 : 0);
